docs: add ADR to adopt ShellCheck (#60)
openedx/.github: docs/decisions/0001-shellcheck.rst

Adopt ShellCheck for shell script linting
#########################################

> Background
> **********
> 
> **Accepted**
> 
> Follow-up issues/PRs:
> 
> * `Add ShellCheck to edx-platform <https://github.com/openedx/edx-platform/pull/31809>`_

Author: ```suggestion
Status
*****

**Accepted**

Initial implementation PRs:

* https://github.com/openedx/.github/pull/64
* https://github.com/openedx/edx-platform/pull/31809
```

* Its error codes, including suggested resolution steps, are `documented on its wiki <https://www.shellcheck.net/wiki/SC1000>`_.
* It can be installed in one line from various `common package managers <https://github.com/koalaman/shellcheck#user-content-installing>`_ including apt and homebrew.
* It can be integrated with `common editors <https://github.com/koalaman/shellcheck#user-content-in-your-editor>`_ including Vim and VSCode.
* As of the writing of this ADR, it `seems to be actively maintained <https://github.com/koalaman/shellcheck/commits/master>`_.

Finally, as a personal anecdote: ShellCheck has been extremely helpful to this ADR's author in `rewriting the edx-platform asset build in Bash <https://github.com/openedx/edx-platform/pull/31791>`_. In addition to catching simple mistakes like typos, ShellCheck's detailed error pages have provided understanding of important shell concepts such as `globbing and word splitting <https://www.shellcheck.net/wiki/SC2086>`_.

Decision
********

Open edX developers are encouraged to use ShellCheck as part of the CI suite for repositories containing shell scripts.

Consequences
************

> * A configurable ShellCheck workflow template will be added to this repository. This will let developers add ShellChecking to their repository's PRs with a couple clicks.

Author: There will still be templates. They will either point to (a) a reusable workflow defined in this repo or (b) an off-the-shelf action. I think I would like to leave the a vs. b question as an [implementation](https://github.com/openedx/.github/pull/64) detail.

> * A configurable ShellCheck workflow template will be added to this repository. This will let developers add ShellChecking to their repository's PRs with a couple clicks.

Reviewer: Ah, I see. Hmm. Not sure if we want to keep the mention (in the next para) of pinning ShellCheck itself, but generally this looks fine.

> * A configurable ShellCheck workflow template will be added to this repository. This will let developers add ShellChecking to their repository's PRs with a couple clicks.

Author: Good point, see https://github.com/openedx/.github/pull/60/files#r1122008719

>   * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introcuded ShellCheck warnings are not expected to be critical.

Reviewer: [nit] typo
```suggestion
  * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introduced ShellCheck warnings are not expected to be critical.
```

>   * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introcuded ShellCheck warnings are not expected to be critical.

Reviewer 2: ```suggestion
  * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck may stick with the default provided in this repo or can override to a different version.  Those that override it  will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introcuded ShellCheck warnings are not expected to be critical.
```

I think letting people default to a thing that automatically upgrades and they only have to think about it when it breaks is nice though I could see an argument against it in favor of being explicit and making it obvious what broke a build.  It would be surprising if there was a random shell error in a PR that was obviously unrelated but I also suspect it would be pretty easy to fix forward given the feature-completeness of shell.

Author: Good suggestion 👍🏻

>   * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck may stick with the default provided in this repo or can override to a different version.  Those that override it  will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introcuded ShellCheck warnings are not expected to be critical.

Author: ```suggestion
  * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck may stick with the default provided in this repo or can override to a different version.  Those that override it  will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introduced ShellCheck warnings are not expected to be critical.
```

>   * In order to ensure deterministic builds, the template will pin ShellCheck to a specific version (v0.9.0 at time of writing). The maintainer of this repository will strive to update the template's version pin over time. Maintainers of repositories using ShellCheck may stick with the default provided in this repo or can override to a different version.  Those that override it  will be encouraged to update their pin over time. We expect that not all maintainers will do this, leading to lag & drift between versions; we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introduced ShellCheck warnings are not expected to be critical.

Author: ```suggestion
  * To support deterministic builds, the template will allow repositories to independently pin ShellCheck to a specific version. We will encourage maintainers of these repositories to update their pin over time. We expect that this will lead to some lag & drift between versions; however, we expect that this will be OK, given that Bash itself is essentially feature-complete, so any newly-introduced ShellCheck warnings are not expected to be critical.
```

@feanil @timmc-edx I'd like to reword this bullet point a bit to make it less specific. That should give us the flexibility over in https://github.com/openedx/.github/pull/64 to decide what combination of template, reusable workflow, and third-party action we want. Let me know if you have any issue with this wording, otherwise I'll merge tomorrow.

Reviewer 2: This seems good to me.

Reviewer: Yep, looks good.

* The new template and this ADR will be announced on the forums and Slack.

* ShellCheck will continue to be used in `openedx-atlas <https://github.com/openedx/openedx-atlas>`_.

* The new ShellCheck template will be used to add checking to edx-platform PRs and master. Existing violations will be resolved or granted amnesty (that is: disabled with a cautionary comment).

> * We will share this ADR with the Maintainership Pilot team, which has been compiling a tenative list of repository standards, so that they may consider ShellCheck as a standard check for applicable repositories.

Reviewer: [nit] typo
```suggestion
* We will share this ADR with the Maintainership Pilot team, which has been compiling a tentative list of repository standards, so that they may consider ShellCheck as a standard check for applicable repositories.
```

Rejected Alternatives
*********************

* **Retroactive application:** We do not plan to retroactively apply ShellCheck to every repository with shell scripts. This would require resolving or applying amnesty to all existing violations, which does not seems worthwhile at this time.

> * **Alternative linting tools:** We did not find any ShellCheck-like tools with comparable documtation, popularity, or ubiquity.

Reviewer: [nit] typo
```suggestion
* **Alternative linting tools:** We did not find any ShellCheck-like tools with comparable documentation, popularity, or ubiquity.
```

> * **ShellCheck in the Python cookiecutters:** We could add the ShellCheck workflow to edx-cookiecutters so that it is run on PRs for all new repositories. However, most new repositories will not need shell scripts, so we do not plan doing this.

Reviewer: [nit] missing word
```suggestion
* **ShellCheck in the Python cookiecutters:** We could add the ShellCheck workflow to edx-cookiecutters so that it is run on PRs for all new repositories. However, most new repositories will not need shell scripts, so we do not plan on doing this.
```
